libkbfs/block_retrieval_worker.go

```diff
@@ -0,0 +1,105 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import (
+	"io"
+	"reflect"
+
+	"golang.org/x/net/context"
+)
+
+// blockRetrievalWorker processes blockRetrievalQueue requests
+type blockRetrievalWorker struct {
+	blockGetter
+	stopCh chan struct{}
+	queue  *blockRetrievalQueue
+}
+
+// run runs the worker loop until Shutdown is called
+func (brw *blockRetrievalWorker) run() {
+	for {
+		select {
+		case <-brw.stopCh:
+			return
+		default:
+			brw.HandleRequest()
+		}
+	}
+}
+
+// newBlockRetrievalWorker returns a blockRetrievalWorker for a given
+// blockRetrievalQueue, using the passed in blockGetter to obtain blocks for
+// requests.
+func newBlockRetrievalWorker(bg blockGetter, q *blockRetrievalQueue) *blockRetrievalWorker {
+	brw := &blockRetrievalWorker{
+		blockGetter: bg,
+		stopCh:      make(chan struct{}),
+		queue:       q,
+	}
+	go brw.run()
+	return brw
+}
+
+// notifyBlockRequestor copies the source block into the request's block
+// pointer, and notifies the channel that the request is waiting on. Should be
+// called in a goroutine.
+func notifyBlockRequestor(req *blockRetrievalRequest, source reflect.Value, err error) {
+	// Copy the decrypted block to the caller
+	dest := reflect.ValueOf(req.block).Elem()
```

Comment:
Actually I can use `requests[0].block`, and then iterate over `requests[1:]` instead of `requests`. Still pretty clean. I like the `NewEmpty` approach though. I'll try to refactor to that.